Cyclic dependency between solana and solana-program-library makes it hard to compile project with personal solana repo

#### Problem

Compiling a repository with a local or solana git repo is not possible because there is cyclic dependency between solana and solana-program-library. 

Here is an example: 

I have a project which uses some solana libraries like `solana-geyser-plugin-interface`, `solana-transaction-status` and few more, and i want to compile it against my personal repository.

```
# cargo file
solana-geyser-plugin-interface = { path = "../solana/geyser-plugin-interface" }
solana-logger = { path = "../solana/solana/logger" }
solana-sdk = { path = "../solana/mangolana/solana/sdk" }
solana-transaction-status = { path = "../solana/transaction-status" }
```
 When i do cargo build I get following error :
 
 ```
error[E0308]: mismatched types
    --> /home/galactus/mangolana/solana/account-decoder/src/parse_account_data.rs:37:18
     |
27   |         m.insert(
     |         - ... which causes `m` to have type `HashMap<solana_sdk::pubkey::Pubkey, ParsableAccount>`
28   |             *ADDRESS_LOOKUP_PROGRAM_ID,
     |             -------------------------- this argument has type `solana_sdk::pubkey::Pubkey`...
...
37   |         m.insert(spl_token::id(), ParsableAccount::SplToken);
     |           ------ ^^^^^^^^^^^^^^^ expected `solana_sdk::pubkey::Pubkey`, found `spl_token_2022::solana_program::pubkey::Pubkey`
     |           |
     |           arguments to this method are incorrect
     |
     = note: `spl_token_2022::solana_program::pubkey::Pubkey` and `solana_sdk::pubkey::Pubkey` have similar names, but are actually distinct types
note: `spl_token_2022::solana_program::pubkey::Pubkey` is defined in crate `solana_program`
    --> /home/galactus/.cargo/registry/src/index.crates.io-6f17d22bba15001f/solana-program-1.16.13/src/pubkey.rs:87:1
     |
87   | pub struct Pubkey(pub(crate) [u8; 32]);
     | ^^^^^^^^^^^^^^^^^
note: `solana_sdk::pubkey::Pubkey` is defined in crate `solana_program`
    --> /home/galactus/mangolana/solana/sdk/program/src/pubkey.rs:87:1
     |
87   | pub struct Pubkey(pub(crate) [u8; 32]);
     | ^^^^^^^^^^^^^^^^^
     = note: perhaps two different versions of crate `solana_program` are being used?
note: method defined here
    --> /home/galactus/.rustup/toolchains/1.70.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/collections/hash/map.rs:1103:12
     |
1103 |     pub fn insert(&mut self, k: K, v: V) -> Option<V> {
     |            ^^^^^^

error[E0308]: mismatched types
    --> /home/galactus/mangolana/solana/account-decoder/src/parse_account_data.rs:38:18
     |
27   |         m.insert(
     |         - ... which causes `m` to have type `HashMap<solana_sdk::pubkey::Pubkey, ParsableAccount>`
28   |             *ADDRESS_LOOKUP_PROGRAM_ID,
     |             -------------------------- this argument has type `solana_sdk::pubkey::Pubkey`...
...
38   |         m.insert(spl_token_2022::id(), ParsableAccount::SplToken2022);
     |           ------ ^^^^^^^^^^^^^^^^^^^^ expected `solana_sdk::pubkey::Pubkey`, found `spl_token_2022::solana_program::pubkey::Pubkey`
     |           |
     |           arguments to this method are incorrect
     |
     = note: `spl_token_2022::solana_program::pubkey::Pubkey` and `solana_sdk::pubkey::Pubkey` have similar names, but are actually distinct types
note: `spl_token_2022::solana_program::pubkey::Pubkey` is defined in crate `solana_program`
    --> /home/galactus/.cargo/registry/src/index.crates.io-6f17d22bba15001f/solana-program-1.16.13/src/pubkey.rs:87:1
     |
87   | pub struct Pubkey(pub(crate) [u8; 32]);
     | ^^^^^^^^^^^^^^^^^
note: `solana_sdk::pubkey::Pubkey` is defined in crate `solana_program`
    --> /home/galactus/mangolana/solana/sdk/program/src/pubkey.rs:87:1
     |
87   | pub struct Pubkey(pub(crate) [u8; 32]);
     | ^^^^^^^^^^^^^^^^^
     = note: perhaps two different versions of crate `solana_program` are being used?
note: method defined here
    --> /home/galactus/.rustup/toolchains/1.70.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/collections/hash/map.rs:1103:12
     |
1103 |     pub fn insert(&mut self, k: K, v: V) -> Option<V> {
     |            ^^^^^^

...
 ```

Here I have no dependency with `spl-token` or `spl-token-2022`.

This makes it hard to compile and test new solana features.

#### Proposed Solution

It am not against this cyclic dependency between both the projects but I believe that definitions that are in solana should stay in solana like `Pubkey`, `Instruction` etc.

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -454,7 +454,9 @@ crossbeam-epoch = { git = "https://github.com/solana-labs/crossbeam", rev = "fd2
 # If you are developing downstream using non-crates-io solana-program (local or
 # forked repo, or from github rev, eg), duplicate the following patch statements
 # in your Cargo.toml. If you still hit duplicate-type errors with the patch
-# statements in place, refresh your Cargo.lock file.
+# statements in place, run `cargo update -p solana-program` and/or `cargo update
+# -p solana-zk-token-sdk` to remove extraneous versions from your Cargo.lock
+# file.
 #
 # There is a similar override in `programs/sbf/Cargo.toml`.  Please keep both comments and the
 # overrides in sync.

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -451,6 +451,11 @@ crossbeam-epoch = { git = "https://github.com/solana-labs/crossbeam", rev = "fd2
 # and we end up with two versions of `solana-program` and `solana-zk-token-sdk` and all of their
 # dependencies in our build tree.
 #
+# If you are developing downstream using non-crates-io solana-program (local or
+# forked repo, or from github rev, eg), duplicate the following patch statements
+# in your Cargo.toml. If you still hit duplicate-type errors with the patch
+# statements in place, refresh your Cargo.lock file.
+#
 # There is a similar override in `programs/sbf/Cargo.toml`.  Please keep both comments and the
 # overrides in sync.
 solana-program = { path = "sdk/program" }

--- a/programs/sbf/Cargo.toml
+++ b/programs/sbf/Cargo.toml
@@ -196,6 +196,11 @@ targets = ["x86_64-unknown-linux-gnu"]
 # and we end up with two versions of `solana-program` and `solana-zk-token-sdk` and all of their
 # dependencies in our build tree.
 #
+# If you are developing downstream using non-crates-io solana-program (local or
+# forked repo, or from github rev, eg), duplicate the following patch statements
+# in your Cargo.toml. If you still hit duplicate-type errors with the patch
+# statements in place, refresh your Cargo.lock file.
+#
 # There is a similar override in `../../Cargo.toml`.  Please keep both comments and the
 # overrides in sync.
 solana-program = { path = "../../sdk/program" }

--- a/programs/sbf/Cargo.toml
+++ b/programs/sbf/Cargo.toml
@@ -199,7 +199,9 @@ targets = ["x86_64-unknown-linux-gnu"]
 # If you are developing downstream using non-crates-io solana-program (local or
 # forked repo, or from github rev, eg), duplicate the following patch statements
 # in your Cargo.toml. If you still hit duplicate-type errors with the patch
-# statements in place, refresh your Cargo.lock file.
+# statements in place, run `cargo update -p solana-program` and/or `cargo update
+# -p solana-zk-token-sdk` to remove extraneous versions from your Cargo.lock
+# file.
 #
 # There is a similar override in `../../Cargo.toml`.  Please keep both comments and the
 # overrides in sync.

